sql/group.go

```diff
@@ -259,93 +316,135 @@ func (v *extractAggregatesVisitor) Visit(expr parser.Expr, pre bool) (parser.Vis
-}
 
-func checkAggregateExpr(expr parser.Expr) error {
-	v := checkAggregateVisitor{}
-	_ = parser.WalkExpr(&v, expr)
-	return v.err
-}
+	if t, ok := expr.(*qvalue); ok {
+		if _, ok := v.groupStrs[t.String()]; ok {
+			return nil, expr
+		}
+		v.aggrErr = fmt.Errorf("column \"%s\" must appear in the GROUP BY clause or be used in an aggregate function", t.col.Name)
+		return v, expr
+	}
 
-type aggregateValue struct {
-	datum parser.Datum
-	expr  *parser.FuncExpr
+	if _, ok := v.groupStrs[expr.String()]; ok {
+		return nil, expr
+	}
+	return v, expr
 }
 
-var _ parser.VariableExpr = &aggregateValue{}
-
-func (*aggregateValue) Variable() {}
+// Check if expressions use aggregation and, if so, if they are valid.
+// "Valid" expressions must either contain no unaggregated qvalues
+// or must appear, verbatim, in the group-by clause. Expressions are
+// string-compared to the group-by clauses as (an approximation of) a
+// recursive expression-tree equality check.
+// Invalid: `SELECT k, SUM(v) FROM kv`
+// - `k` is unaggregated and does not appear in the (missing) GROUP BY.
+// Valid: `SELECT k, SUM(v) FROM kv GROUP BY k`
+// Also valid: `SELECT UPPER(k), SUM(v) FROM kv GROUP BY UPPER(k)`
+// - `UPPER(k)` appears in GROUP BY.
+// Also valid: `SELECT UPPER(k), SUM(v) FROM kv GROUP BY k`
+// - `k` appears in GROUP BY, so `UPPER(k)` is OK, but...
+// Invalid: `SELECT k, SUM(v) FROM kv GROUP BY UPPER(k)`
+// - qvalue subtrees from the select must appear *verbatim* in the GROUP BY.
+func checkAggregateExprs(group parser.GroupBy, exprs []parser.Expr) (bool, error) {
+	aggregated := len(group) > 0
 
-func (av *aggregateValue) String() string {
-	return av.expr.String()
-}
+	v := checkAggregateVisitor{}
 
-func (av *aggregateValue) Walk(v parser.Visitor) {
-	// I expected to implement:
-	// av.datum = parser.WalkExpr(v, av.datum).(parser.Datum)
-	// But it seems `av.datum` is sometimes nil.
-}
+	// TODO(dt): consider other ways of comparing expression trees.
+	v.groupStrs = make(map[string]struct{}, len(group))
+	for i := range group {
+		v.groupStrs[group[i].String()] = struct{}{}
+	}
 
-func (av *aggregateValue) TypeCheck(args parser.MapArgs) (parser.Datum, error) {
-	return av.expr.TypeCheck(args)
+	for _, expr := range exprs {
+		_ = parser.WalkExpr(&v, expr)
+		if v.aggregated {
+			aggregated = true
+		}
+		if v.aggrErr != nil && aggregated {
+			return aggregated, v.aggrErr
+		}
+	}
+	return aggregated, nil
 }
 
-func (av *aggregateValue) Eval(ctx parser.EvalContext) (parser.Datum, error) {
-	return av.datum.Eval(ctx)
-}
+var _ parser.VariableExpr = &aggregateFunc{}
 
 type aggregateFunc struct {
-	val  aggregateValue
-	impl aggregateImpl
-	seen map[string]struct{}
+	expr    parser.Expr
+	arg     parser.Expr
+	create  func() aggregateImpl
+	group   *groupNode
+	buckets map[string]aggregateImpl
+	seen    map[string]struct{}
 }
 
-func (a *aggregateFunc) Add(d parser.Datum) error {
+func (a *aggregateFunc) add(bucket string, d parser.Datum) error {
```

Comment:
so, it is only converted back to a `[]byte` for the `seen` distinct case -- and it _is_ used as a string (since map keys can't be []byte, right?). deferring the `sting()` conversion would mean doing twice -- once here and once in the caller... I guess we could just pass both? ugh.